### PR TITLE
Load Certificates from $SSL_CERT_DIR

### PR DESCRIPTION
**Pull request by nebhale**
Previously, when the the CertificateLoader loaded, it would parse a single magical file for certificates.  To support the customization of CA certificates though, it should have been honoring the $SSL_CERT_DIR path-delimited environment variable for the certificate locations.

This change adds support for parsing the contents of this environment variable, falling back to the Ubuntu default when it is not set.

### CODE REVIEW

**Comment by nebhale** [comment]
I've done local validation that this works in the downstream buildpacks and that it does not add any significant overhead.

File: certificate_loader.go [hunk 1]
```diff
@@ -22,19 +22,21 @@ import (
...
 
-const CACertificates = "/etc/ssl/certs/ca-certificates.crt"
+const DefaultCertsDir = "/etc/ssl/certs"
 
 type CertificateLoader struct {
-	CACertificatesPath string
-	KeyStorePassword   string
-	KeyStorePath       string
-	Logger             io.Writer
+	CertificateDirs  []string
```
Comment (ekcasey) [inline certificate_loader.go:36]: Digging into the openssl docs, it seems that we should support the case where both `SSL_CERT_FILE` and `SSL_CERT_DIR` are set.

https://www.openssl.org/docs/man1.1.0/man3/SSL_CTX_set_default_verify_paths.html

**Review by ekcasey — CHANGES_REQUESTED** [review]

File: certificate_loader.go [hunk 2]
```diff
@@ -74,24 +76,48 @@ func (c *CertificateLoader) Load() error {
...
 
-	var (
-		block  *pem.Block
-		blocks []*pem.Block
-	)
-	for len(rest) != 0 {
-		block, rest = pem.Decode(rest)
-		blocks = append(blocks, block)
+		for _, c := range children {
+			if c.IsDir() || !re.MatchString(c.Name()) {
```
[the remark this answers is not in the document]
Comment (nebhale) [reply certificate_loader.go:92]: Going to skip this as it appears to be in conflict with OpenSSL's defined algorithm.

File: certificate_loader.go [hunk 3]
```diff
@@ -17,113 +17,192 @@
...
 
-	if err := keystore.Encode(out, ks, []byte(c.KeyStorePassword)); err != nil {
+	if err := keystore.Encode(out, ks, []byte(password)); err != nil {
 		return fmt.Errorf("unable to encode keystore\n%w", err)
 	}
 
 	return nil
 }
+
+func CertificateLoaderMetadata(certificateLoader CertificateLoader) (map[string]interface{}, error) {
```
Comment (ekcasey) [inline certificate_loader.go:185]: Any reason this isn't a method on the `CertificateLoader`?

**Review by ekcasey — APPROVED** [review]

File: certificate_loader.go [hunk 4]
```diff
@@ -17,113 +17,192 @@
...
-	if unix.Access(c.KeyStorePath, unix.W_OK) != nil {
-		_, _ = fmt.Fprintf(c.Logger, "WARNING: Unable to add container CA certificates to JVM because %s is read-only", c.KeyStorePath)
+func (c CertificateLoader) writeKeyStore(ks keystore.KeyStore, path string, password string) error {
+	if unix.Access(path, unix.W_OK) != nil {
+		_, _ = fmt.Fprintf(c.Logger, "WARNING: Unable to add container CA certificates to JVM because %s is read-only", path)
 		return nil
 	}
 
-	out, err := os.OpenFile(c.KeyStorePath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0644)
+	out, err := os.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0644)
```
[the remark this answers is not in the document]
Comment (nebhale) [reply certificate_loader.go:172]: This is a security file, much like SSH artifacts, and therefore being explicit about `0644` is the appropriate behavior.

File: jdk.go [hunk 5]
```diff
@@ -31,33 +28,30 @@ import (
 )
 
 type JDK struct {
-	Certificates     string
-	LayerContributor libpak.DependencyLayerContributor
-	Logger           bard.Logger
+	CertificateLoader CertificateLoader
```
[the remark this answers is not in the document]
Comment (nebhale) [reply jdk.go:31]: What would that interface be named?

File: certificate_loader.go (hunk 3 above)
Comment (nebhale) [reply certificate_loader.go:185]: A composition style, but I don't have a strong opinion either way.

**Pull request merged**